recursion/compiler/src/circuit/mod.rs

```diff
@@ -0,0 +1,606 @@
+            // DslIr::IfEq(_, _, _, _) => todo!(),
+            // DslIr::IfNe(_, _, _, _) => todo!(),
+            // DslIr::IfEqI(_, _, _, _) => todo!(),
+            // DslIr::IfNeI(_, _, _, _) => todo!(),
+            // DslIr::Break => todo!(),
+            // DslIr::Alloc(_, _, _) => todo!(),
+            // DslIr::LoadV(_, _, _) => todo!(),
+            // DslIr::LoadF(_, _, _) => todo!(),
+            // DslIr::LoadE(_, _, _) => todo!(),
+            // DslIr::StoreV(_, _, _) => todo!(),
+            // DslIr::StoreF(_, _, _) => todo!(),
+            // DslIr::StoreE(_, _, _) => todo!(),
+            // DslIr::CircuitNum2BitsV(_, _, _) => todo!(),
+            // DslIr::CircuitNum2BitsF(_, _) => todo!(),
+            // DslIr::Poseidon2PermuteBabyBear(_, _) => todo!(),
+            // DslIr::Poseidon2CompressBabyBear(_, _, _) => todo!(),
+            // DslIr::Poseidon2AbsorbBabyBear(_, _) => todo!(),
+            // DslIr::Poseidon2FinalizeBabyBear(_, _) => todo!(),
+            // DslIr::CircuitPoseidon2Permute(_) => todo!(),
+            // DslIr::CircuitPoseidon2PermuteBabyBear(_) => todo!(),
+            // DslIr::HintBitsU(_, _) => todo!(),
+            // DslIr::HintBitsV(_, _) => todo!(),
+            // DslIr::HintBitsF(_, _) => todo!(),
+            // DslIr::PrintV(_) => todo!(),
+            // DslIr::PrintF(_) => todo!(),
+            // DslIr::PrintE(_) => todo!(),
+            // DslIr::Error() => todo!(),
+            // DslIr::HintExt2Felt(_, _) => todo!(),
+            // DslIr::HintLen(_) => todo!(),
+            // DslIr::HintVars(_) => todo!(),
+            // DslIr::HintFelts(_) => todo!(),
+            // DslIr::HintExts(_) => todo!(),
+            // DslIr::WitnessVar(_, _) => todo!(),
+            // DslIr::WitnessFelt(_, _) => todo!(),
+            // DslIr::WitnessExt(_, _) => todo!(),
+            // DslIr::Commit(_, _) => todo!(),
+            // DslIr::RegisterPublicValue(_) => todo!(),
+            // DslIr::Halt => todo!(),
+            // DslIr::CircuitCommitVkeyHash(_) => todo!(),
+            // DslIr::CircuitCommitCommitedValuesDigest(_) => todo!(),
+            // DslIr::FriFold(_, _) => todo!(),
+            // DslIr::CircuitSelectV(_, _, _, _) => todo!(),
+            // DslIr::CircuitSelectF(_, _, _, _) => todo!(),
+            // DslIr::CircuitSelectE(_, _, _, _) => todo!(),
+            // DslIr::CircuitExt2Felt(_, _) => todo!(),
+            // DslIr::CircuitFelts2Ext(_, _) => todo!(),
+            // DslIr::LessThan(_, _, _) => todo!(),
+            // DslIr::CycleTracker(_) => todo!(),
+            // DslIr::ExpReverseBitsLen(_, _, _) => todo!(),
+            instr => panic!("unsupported instruction: {instr:?}"),
+        }
+    }
+
+    /// Emit the instructions from a list of operations in the DSL.
+    pub fn compile(
+        &mut self,
+        operations: TracedVec<DslIr<AsmConfig<F, EF>>>,
+    ) -> Vec<Instruction<F>> {
+        // First, preprocess.
```

Comment:
Don't write comments like this, where the line breaks are random. if you want to write a step-by-step list, use a bullet point list or "-".